File: frontend/src/types/placement.ts
export type FilingStatus = 'single' | 'married_joint';
export type AccountCharacter = 'sheltered' | 'taxable';
export type Provenance = 'fetched' | 'derived' | 'stated' | 'unsourced';

/** Every assumption field carries its provenance. `unsourced` renders flagged. */
export interface SourcedFloat {
  value: number;
  source: Provenance;
  note: string;
}

export interface FundAssumption {
  ticker: string;
  gross_yield: SourcedFloat;
  qualified_pct: SourcedFloat;
  foreign_share: SourcedFloat;
  region: string;
}

export interface RegionWithholding {
  region: string;
  rate: SourcedFloat;
}

export interface AccountContribution {
  name: string;
  amount: number;
  character: AccountCharacter;
}

export interface TaxInputs {
  gross_comp: number;
  filing_status: FilingStatus;
  state_rate: number;
  pre_tax_deferral: number;
  ftc_recovery: number;
}

export interface GrowthInputs {
  years: number;
  real_return: number;
}

export interface PlacementRequest {
  tax: TaxInputs;
  growth: GrowthInputs;
  funds: FundAssumption[];
  withholding: RegionWithholding[];
  accounts: AccountContribution[];
  target_split: Record<string, number>;
}

export interface MarginalRates {
  taxable_income: number;
  standard_deduction: number;
  fed_ordinary: number;
  fed_ltcg: number;
  niit: number;
  state_rate: number;
  ordinary_rate: number;
  qualified_rate: number;
}

export interface FundDrag {
  ticker: string;
  gross_yield: number;
  qualified_pct: number;
  foreign_share: number;
  withholding_rate: number;
  /** phi = q * foreign_share * withholding */
  foreign_tax: number;
  /** kappa*qual + (1-kappa)*ord */
  blended_rate: number;
  /** q * blended */
  us_tax: number;
  drag_taxable: number;
  drag_sheltered: number;
  benefit: number;
}

export interface AllocationCell {
  account: string;
  ticker: string;
  amount: number;
}

export interface Strategy {
  name: string;
  cells: AllocationCell[];
  sheltered: Record<string, number>;
  annual_drag: number;
  drag_bps: number;
}

/** Two independent routes to the saving; they must agree. */
export interface SavingCheck {
  direct: number;
  decomposed: number;
  agree: boolean;
}

export interface GrowthPoint {
  year: number;
  vanilla_total: number;
  optimal_total: number;
  gap: number;
}

export interface PlacementResponse {
  rates: MarginalRates;
  drags: FundDrag[];
  total_contributions: number;
  sheltered_capacity: number;
  taxable_capacity: number;
  sheltered_share: number;
  fund_targets: Record<string, number>;
  vanilla: Strategy;
  optimal: Strategy;
  annual_saving: number;
  saving_bps: number;
  saving_check: SavingCheck;
  growth: GrowthPoint[];
  vanilla_terminal: number;
  optimal_terminal: number;
  terminal_gap: number;
  warnings: string[];
}

export interface ReferenceYield {
  ticker: string;
  ttm_yield: number | null;
  price: number | null;
  /** 'live' | 'cache' | 'unavailable' */
  source: string;
  as_of: string | null;
}

export interface DefaultsResponse {
  request: PlacementRequest;
  reference_yields: ReferenceYield[];
}

File: backend/app/schemas/placement.py
from typing import Dict, List, Optional

from pydantic import BaseModel

from ..core.placement_models import PlacementRequest
from ..services.yield_reference import ReferenceYield


class MarginalRatesOut(BaseModel):
    """Derived from gross compensation, never asked for directly."""
    taxable_income: float
    standard_deduction: float
    fed_ordinary: float
    fed_ltcg: float
    niit: float
    state_rate: float
    ordinary_rate: float
    qualified_rate: float


class FundDragOut(BaseModel):
    """One row of the Panel 1 matrix, with the decomposition kept inline.

    Every intermediate is returned so each cell is auditable against the
    formulas rather than taken on trust.
    """
    ticker: str
    gross_yield: float
    qualified_pct: float
    foreign_share: float
    withholding_rate: float
    # phi = q * foreign_share * withholding
    foreign_tax: float
    # kappa*qual + (1-kappa)*ord
    blended_rate: float
    # q * blended
    us_tax: float
    drag_taxable: float
    drag_sheltered: float
    benefit: float


class AllocationCell(BaseModel):
    account: str
    ticker: str
    amount: float


class StrategyOut(BaseModel):
    name: str
    cells: List[AllocationCell]
    # ticker -> dollars in the sheltered column
    sheltered: Dict[str, float]
    annual_drag: float
    drag_bps: float


class SavingCheck(BaseModel):
    """Two independent routes to the same number; they must agree.

    `direct` is vanilla drag minus optimal drag. `decomposed` is
    sum(benefit_i * (sheltered_optimal_i - sheltered_vanilla_i)).
    """
    direct: float
    decomposed: float
    agree: bool


class GrowthPoint(BaseModel):
    year: int
    vanilla_total: float
    optimal_total: float
    gap: float


class PlacementResponse(BaseModel):
    rates: MarginalRatesOut
    drags: List[FundDragOut]

    total_contributions: float
    sheltered_capacity: float
    taxable_capacity: float
    sheltered_share: float
    fund_targets: Dict[str, float]

    vanilla: StrategyOut
    optimal: StrategyOut
    annual_saving: float
    saving_bps: float
    saving_check: SavingCheck

    growth: List[GrowthPoint]
    vanilla_terminal: float
    optimal_terminal: float
    terminal_gap: float

    # Non-blocking input problems, e.g. a target split that does not sum to 1.
    warnings: List[str]


class DefaultsResponse(BaseModel):
    request: PlacementRequest
    reference_yields: List[ReferenceYield]


class SolveRequest(BaseModel):
    request: PlacementRequest


class ReferenceYieldsResponse(BaseModel):
    reference_yields: List[ReferenceYield]
    as_of: Optional[str] = None
